### DISCUSSION TITLE
Duckdb for external database

### DISCUSSION THREAD

**Discussion by sebastian9610** [duckdb/duckdb#13011] [Q&A]
Hello, I want to know if it is possible to connect with duckdb to an external database such as SQL Server or Redshift to save the query result directly to the duckdb database. Currently I do it first using pandas to extract the information but I want to skip this step due to excessive memory usage

Example

`import pandas as pd
import duckdb

cursor = duckdb.connect('mydatabase.db')
cursor.execute('CREATE SCHEMA CCC;')

df_table = pd.read_sql("""SELECT COL1, COL2
    FROM schema.ventas
    WHERE COL3 IN ('a', 'b', 'c');""", con=cnxn)

cursor.execute("""CREATE TABLE CCC.MyTable AS SELECT t1.*, t2.customers
    FROM df_table t1
    LEFT JOIN read_csv('customers.csv') t2 ON t1.COL1 = t2.COL1;""")`

**Comment by FrancoisLepoutre** [comment]
Hi,
I have no specific answer for this question as I have never been a pandas user. But sure **_it would make sense to avoid using a memory-constrained dataframe technology to act as a middleware to feed our great duck_**. 

As a generic pattern, one could propose to use db-centric middleware to automate linking between database technologies. Currently **_sqlserver and duckdb both handle the "old hat" ODBC_**, historically the first generic inter-database communication protocol.

ODBC is pretty easy to handle on MS-Windows and a bit more hacky on linux boxes. But it may possibly fit your needs. Should this works, expect speed to be a couple of generations backwards but memory-consumption to be low and, possibly, stability to be quite good. Should you run some test, I would enjoy reading your feedback here by the way!

PS: there are of course numerous alternatives middleware for feeding such as _Apache Arrow_ and more.

**Reply by pietercrous** [reply]
Are there any examples of connecting to ODBC/JDBC sources in DuckDB (Python), where I can attach the database to the DuckDB instance? Does the ADBC driver work with MS SQL Server yet?

**Reply by Hugoberry** [reply]
I've developed a cross-platform ODBC extension for this purpose: nanodbc. I've tested it with SQL Server, so it should also work for your Redshift setup. Would be interested to hear any feedback https://github.com/Hugoberry/duckdb-nanodbc-extension

**Comment by NickCrews** [comment]
This doesn't exactly answer your question, but [ibis](https://ibis-project.org/) is good for this sort of moving data between databases. Read from your source, materialize to arrow or to disk depending on the size, and then ingest into the destination.

**Comment by gustavoleo** [comment]
> Hello, I want to know if it is possible to connect with duckdb to an external database such as SQL Server or Redshift to save the query result directly to the duckdb database. Currently I do it first using pandas to extract the information but I want to skip this step due to excessive memory usage
> 
> 
> 
> Example
> 
> 
> 
> `import pandas as pd
> 
> import duckdb
> 
> 
> 
> cursor = duckdb.connect('mydatabase.db')
> 
> cursor.execute('CREATE SCHEMA CCC;')
> 
> 
> 
> df_table = pd.read_sql("""SELECT COL1, COL2
> 
>     FROM schema.ventas
> 
>     WHERE COL3 IN ('a', 'b', 'c');""", con=cnxn)
> 
> 
> 
> cursor.execute("""CREATE TABLE CCC.MyTable AS SELECT t1.*, t2.customers
> 
>     FROM df_table t1
> 
>     LEFT JOIN read_csv('customers.csv') t2 ON t1.COL1 = t2.COL1;""")`

You should try connector-x and experience the speed of the light
